### PR TITLE
r/mq_broker - allow updating engine version

### LINKED ISSUE #11827: Changing aws_mq_broker engine_version should not force resource replacement
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Amazon MQ allows you to upgrade the engine version in place, however the AWS provider forces a replacement of the resource if the engine_version is updated.

### New or Affected Resource(s)

* aws_mq_broker

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #11827

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resouce_aws_mq_broker - allow updating engine version
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSMqBroker_updateEngineVersion'
--- PASS: TestAccAWSMqBroker_updateEngineVersion (1394.18s)
```

### CODE REVIEW

**Comment by ewbankkit** [comment]
Verified acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSMqBroker_updateEngineVersion'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAWSMqBroker_updateEngineVersion -timeout 120m
=== RUN   TestAccAWSMqBroker_updateEngineVersion
=== PAUSE TestAccAWSMqBroker_updateEngineVersion
=== CONT  TestAccAWSMqBroker_updateEngineVersion
--- PASS: TestAccAWSMqBroker_updateEngineVersion (1515.99s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1516.030s
```

**Comment by ewbankkit** [comment]
Verified acceptance test:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSMqBroker_updateEngineVersion'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAWSMqBroker_updateEngineVersion -timeout 120m
=== RUN   TestAccAWSMqBroker_updateEngineVersion
=== PAUSE TestAccAWSMqBroker_updateEngineVersion
=== CONT  TestAccAWSMqBroker_updateEngineVersion
--- PASS: TestAccAWSMqBroker_updateEngineVersion (1466.44s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1466.464s
```

**Comment by DrFaust92** [comment]
testing refactor + rebase, test re-run:

```
--- PASS: TestAccAWSMqBroker_allFieldsDefaultVpc (1941.61s)
--- PASS: TestAccAWSMqBroker_allFieldsCustomVpc (2069.22s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_KmsKeyId (1153.99s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled (1159.08s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled (1149.53s)
--- PASS: TestAccAWSMqBroker_updateUsers (1638.19s)
--- PASS: TestAccAWSMqBroker_tags (1195.53s)
--- PASS: TestAccAWSMqBroker_updateSecurityGroup (1420.80s)
--- PASS: TestAccAWSMqBroker_disappears (1209.47s)
--- PASS: TestAccAWSMqBroker_updateEngineVersion (1425.13s)
```

**Comment by szymonpk** [comment]
Any chance we can have this merged?

**Comment by Matiasmct** [comment]
Any chance we can have this merged?

**Comment by YakDriver** [comment]
@DrFaust92 This looks like a very helpful PR. Now that some rework with #16108 is merged, if you have a chance to rebase, it would be great to also get this fix in.

**Comment by DrFaust92** [comment]
Thanks @YakDriver rebased and reran tests:

```
--- PASS: TestAccAWSMqBroker_rabbitmq (541.35s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled (1124.17s)
--- PASS: TestAccAWSMqBroker_basic (1155.99s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_KmsKeyId (1157.90s)
--- PASS: TestAccAWSMqBroker_disappears (1171.16s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled (1200.44s)
--- PASS: TestAccAWSMqBroker_tags (1267.72s)
--- PASS: TestAccAWSMqBroker_updateSecurityGroup (1402.44s)
--- PASS: TestAccAWSMqBroker_updateEngineVersion (1409.84s)
--- PASS: TestAccAWSMqBroker_updateUsers (1594.24s)
--- PASS: TestAccAWSMqBroker_allFieldsDefaultVpc (1772.55s)
--- PASS: TestAccAWSMqBroker_allFieldsCustomVpc (1846.14s)
```

File: .changelog/12758.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+```release-note:enhancement
+resource/aws_mq_broker: Support updating broker
```
Comment (ewbankkit) [inline .changelog/12758.txt:2]:
```suggestion
resource/aws_mq_broker: Support updating broker engine version
```

**Comment by ewbankkit** [comment]
Verified acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSMqBroker_' ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSMqBroker_ -timeout 120m
=== RUN   TestAccAWSMqBroker_basic
=== PAUSE TestAccAWSMqBroker_basic
=== RUN   TestAccAWSMqBroker_allFieldsDefaultVpc
=== PAUSE TestAccAWSMqBroker_allFieldsDefaultVpc
=== RUN   TestAccAWSMqBroker_allFieldsCustomVpc
=== PAUSE TestAccAWSMqBroker_allFieldsCustomVpc
=== RUN   TestAccAWSMqBroker_EncryptionOptions_KmsKeyId
=== PAUSE TestAccAWSMqBroker_EncryptionOptions_KmsKeyId
=== RUN   TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled
=== PAUSE TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled
=== RUN   TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled
=== PAUSE TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled
=== RUN   TestAccAWSMqBroker_updateUsers
=== PAUSE TestAccAWSMqBroker_updateUsers
=== RUN   TestAccAWSMqBroker_tags
=== PAUSE TestAccAWSMqBroker_tags
=== RUN   TestAccAWSMqBroker_updateSecurityGroup
=== PAUSE TestAccAWSMqBroker_updateSecurityGroup
=== RUN   TestAccAWSMqBroker_updateEngineVersion
=== PAUSE TestAccAWSMqBroker_updateEngineVersion
=== RUN   TestAccAWSMqBroker_disappears
=== PAUSE TestAccAWSMqBroker_disappears
=== RUN   TestAccAWSMqBroker_rabbitmq
=== PAUSE TestAccAWSMqBroker_rabbitmq
=== CONT  TestAccAWSMqBroker_basic
=== CONT  TestAccAWSMqBroker_tags
=== CONT  TestAccAWSMqBroker_rabbitmq
--- PASS: TestAccAWSMqBroker_basic (1115.06s)
--- PASS: TestAccAWSMqBroker_tags (1178.00s)
=== CONT  TestAccAWSMqBroker_disappears
--- PASS: TestAccAWSMqBroker_rabbitmq (539.40s)
=== CONT  TestAccAWSMqBroker_updateEngineVersion
--- PASS: TestAccAWSMqBroker_disappears (1114.89s)
=== CONT  TestAccAWSMqBroker_updateSecurityGroup
--- PASS: TestAccAWSMqBroker_updateEngineVersion (1287.01s)
=== CONT  TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled
--- PASS: TestAccAWSMqBroker_updateSecurityGroup (1344.09s)
=== CONT  TestAccAWSMqBroker_updateUsers
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled (1075.88s)
=== CONT  TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled
--- PASS: TestAccAWSMqBroker_updateUsers (1464.31s)
=== CONT  TestAccAWSMqBroker_allFieldsCustomVpc
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled (1167.66s)
=== CONT  TestAccAWSMqBroker_EncryptionOptions_KmsKeyId
--- PASS: TestAccAWSMqBroker_EncryptionOptions_KmsKeyId (1149.59s)
=== CONT  TestAccAWSMqBroker_allFieldsDefaultVpc
--- PASS: TestAccAWSMqBroker_allFieldsDefaultVpc (1561.53s)
--- PASS: TestAccAWSMqBroker_allFieldsCustomVpc (1782.41s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	6883.803s
```

**Comment by YakDriver** [comment]
Ut-oh. Looks like a bug in our `waiting-response` logic. It didn't remove the label when you commented 3 days ago. 🤔 

Anyway, thanks for working on this!

File: aws/resource_aws_mq_broker_test.go [hunk 2]
```diff
@@ -889,10 +911,10 @@ func TestAccAWSMqBroker_disappears(t *testing.T) {
 		CheckDestroy: testAccCheckAwsMqBrokerDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccMqBrokerConfig(sgName, brokerName),
+				Config: testAccMqBrokerConfig(rName),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckAwsMqBrokerExists(resourceName, &broker),
-					testAccCheckAwsMqBrokerDisappears(&broker),
+					testAccCheckResourceDisappears(testAccProvider, resourceAwsMqBroker(), resourceName),
```
Comment (YakDriver) [inline aws/resource_aws_mq_broker_test.go:965]: Nice!

File: aws/resource_aws_mq_broker.go [hunk 3]
```diff
@@ -896,13 +895,13 @@ func flattenMqBrokerInstances(instances []*mq.BrokerInstance) []interface{} {
 	for i, instance := range instances {
 		m := make(map[string]interface{})
 		if instance.ConsoleURL != nil {
-			m["console_url"] = *instance.ConsoleURL
+			m["console_url"] = aws.StringValue(instance.ConsoleURL)
```
Comment (YakDriver) [inline aws/resource_aws_mq_broker.go:898]: Good!

File: aws/resource_aws_mq_broker.go [hunk 4]
```diff
@@ -356,7 +354,7 @@ func resourceAwsMqBrokerCreate(d *schema.ResourceData, meta interface{}) error {
 		input.MaintenanceWindowStartTime = expandMqWeeklyStartTime(v.([]interface{}))
 	}
 	if v, ok := d.GetOk("security_groups"); ok && v.(*schema.Set).Len() > 0 {
-		input.SecurityGroups = expandStringSet(d.Get("security_groups").(*schema.Set))
```
Comment (YakDriver) [inline aws/resource_aws_mq_broker.go:359]: Good catch!

**Review by YakDriver — APPROVED** [review]
Looks great! 🎉 

Acceptance tests on commercial (`us-west-2`):
```
--- PASS: TestAccAWSMqBroker_rabbitmq (516.10s)
--- PASS: TestAccAWSMqBroker_throughputOptimized (1015.89s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_KmsKeyId (1100.35s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled (1102.55s)
--- PASS: TestAccAWSMqBroker_tags (1120.97s)
--- PASS: TestAccAWSMqBroker_basic (1131.82s)
--- PASS: TestAccAWSMqBroker_disappears (1138.39s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled (1146.32s)
--- PASS: TestAccAWSMqBroker_updateSecurityGroup (1340.00s)
--- PASS: TestAccAWSMqBroker_updateEngineVersion (1340.28s)
--- PASS: TestAccAWSMqBroker_updateUsers (1507.62s)
--- PASS: TestAccAWSMqBroker_allFieldsDefaultVpc (1797.68s)
--- PASS: TestAccAWSMqBroker_allFieldsCustomVpc (1818.05s)
```

On GovCloud (service not available):
```
--- SKIP: TestAccAWSMqBroker_updateUsers (1.23s)
--- SKIP: TestAccAWSMqBroker_rabbitmq (1.23s)
--- SKIP: TestAccAWSMqBroker_basic (1.23s)
--- SKIP: TestAccAWSMqBroker_allFieldsCustomVpc (1.23s)
--- SKIP: TestAccAWSMqBroker_disappears (1.23s)
--- SKIP: TestAccAWSMqBroker_updateEngineVersion (1.23s)
--- SKIP: TestAccAWSMqBroker_tags (1.23s)
--- SKIP: TestAccAWSMqBroker_updateSecurityGroup (1.23s)
--- SKIP: TestAccAWSMqBroker_throughputOptimized (1.23s)
--- SKIP: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled (1.23s)
--- SKIP: TestAccAWSMqBroker_allFieldsDefaultVpc (1.23s)
--- SKIP: TestAccAWSMqBroker_EncryptionOptions_KmsKeyId (1.23s)
--- SKIP: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled (1.23s)
```

**Pull request merged**